### PR TITLE
init redux example

### PR DESCRIPTION
**Pull request by EvgenyParomov**
Добавление примера написанного на redux, redux-tookit
По большей части это копия примера на effector но все модели переписаны на redux

Главная идея, как сохранить модульность и инкапсуляцию при работе с единым redux состоянием

### CODE REVIEW

File: todo-app-redux/src/app/styles/normalize-antd.scss [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Лоадер на весь контейнер
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/app/styles/normalize-antd.scss:1]: I'm not sure we need this comment here at all, but if you think we do, please use English

File: todo-app-redux/src/entities/task/model/index.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Разбиваем на submodels
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/entities/task/model/index.ts:1]: -> English

File: todo-app-redux/src/entities/task/model/tasks.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  createSlice,
+  createEntityAdapter,
+  PayloadAction,
+  createSelector,
+} from "@reduxjs/toolkit";
+
+import { typicodeApi } from "shared/api";
+import { createBaseSelector } from "shared/lib/redux-std";
+
+// Этот файл можно разделить на меньшие файлы при увеличении в нём колличество логики,
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/entities/task/model/tasks.ts:12]: -> English

File: todo-app-redux/src/entities/task/model/tasks.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+const adapterSelectors = entityAdapter.getSelectors();
+
+const initialState = entityAdapter.getInitialState({
+  isTaskListLoading: false,
+  isTaskDetailsLoading: false,
+});
+
+type State = typeof initialState;
+
+// Повторение структуры папок упрощает навигацию и защищает от конфликтов
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/entities/task/model/tasks.ts:28]: -> English

File: todo-app-redux/src/entities/task/model/tasks.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+const details = createSelector(
+  baseSelector,
+  (_: unknown, id: number) => id,
+  adapterSelectors.selectById
+);
+
+// public API
+
+// Мы определяем только публичные экшоны, мы защищаем инварианты нашего состояния
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/entities/task/model/tasks.ts:110]: -> English

File: todo-app-redux/src/entities/task/model/tasks.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+// (например инвариант что isTaskListLoading === true когда происходит выборка)
+export const actions = {
+  getList,
+  getById,
+  toggleCompleted: slice.actions.toggleCompleted,
+};
+
+// Мы определяем только публичные селекторы, таким образом мы скрываем структуру данных нашего состояния
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/entities/task/model/tasks.ts:118]: -> English

File: todo-app-redux/src/entities/task/model/tasks.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+export const selectors = {
+  isListLoading,
+  isDetailsLoading,
+  list,
+  isListEmpty,
+  details,
+};
+
+// Мы экспортируем структуру редьюссеров целиком, таким образом reducerPath остаётся приватным
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/entities/task/model/tasks.ts:127]: -> English

File: todo-app-redux/src/entities/task/ui/task-row/index.tsx [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    data: import("shared/api").Task;
+    titleHref?: string;
+    before?: ReactNode;
+}>;
+
+export const TaskRow = ({ data, before, titleHref }: TaskRowProps) => {
+    const title = titleHref ? <Link to={titleHref}>{data.title}</Link> : data.title
+
+    return (
+        // Можно смело использовать classnames и аналоги
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/entities/task/ui/task-row/index.tsx:17]:
```suggestion
```

File: todo-app-redux/src/entities/task/ui/task-row/index.tsx [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import type { PropsWithChildren, ReactNode } from "react";
+import cn from "classnames"; // Можно смело использовать аналоги
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/entities/task/ui/task-row/index.tsx:2]:
```suggestion
import cn from "classnames";
```

File: todo-app-redux/src/entities/task/ui/task-card/index.tsx [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    titleHref?: string;
+}> & import("antd").CardProps;
+
+export const TaskCard = ({ data, titleHref, children, ...cardProps }: TaskCardProps) => {
+    // Можно обработать и получше при желании
+    if (!data && !cardProps.loading) return null;
+
+    return (
+        <Card
+            // Можно обработать и получше при желании
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/entities/task/ui/task-card/index.tsx:17]:
```suggestion
```

File: todo-app-redux/src/entities/task/ui/task-card/index.tsx [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+import { Link } from "react-router-dom";
+import styles from "./styles.module.scss";
+
+export type TaskCardProps = PropsWithChildren<{
+    data?: import("shared/api").Task;
+    titleHref?: string;
+}> & import("antd").CardProps;
+
+export const TaskCard = ({ data, titleHref, children, ...cardProps }: TaskCardProps) => {
+    // Можно обработать и получше при желании
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/entities/task/ui/task-card/index.tsx:12]:
```suggestion
```

File: todo-app-redux/src/entities/task/ui/task-row/styles.module.scss [hunk 12]
```diff
@@ -0,0 +1,11 @@
+.root {
+    background: #fff;
+    padding: 20px;
+    transition: 0.25s;
+}
+
+.completed {
+    // Можно и получше подобрать стили
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/entities/task/ui/task-row/styles.module.scss:8]:
```suggestion
```

File: todo-app-redux/src/entities/task/model/tasks.ts [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    return typicodeApi.tasks
+      .getTaskById(params)
+      .then((res) => res.data)
+      .finally(() => {
+        dispatch(slice.actions.setTaskListLoading(false));
+      });
+  }
+);
+
+// selecotrs
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/entities/task/model/tasks.ts:87]:
```suggestion
// selectors
```

File: todo-app-redux/src/features/tasks-filters/config.ts [hunk 14]
```diff
@@ -0,0 +1,32 @@
+import { TaskFilters } from "./model/task-filters.type";
+
+export type Filter = {
+  id: number;
+  title: string;
+  config: TaskFilters;
+};
+
+// Описываем здесь датасет фильтров "Закрытые" / "Открытые" и т.п.
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/features/tasks-filters/config.ts:9]: -> English

File: todo-app-redux/src/pages/index.tsx [hunk 15]
```diff
@@ -0,0 +1,16 @@
+// Либо использовать @loadable/component, в рамках туториала - некритично
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/pages/index.tsx:1]:
```suggestion
```

File: todo-app-redux/src/shared/api/typicode/base.ts [hunk 16]
```diff
@@ -0,0 +1,7 @@
+import axios from "axios";
+import { API_URL } from "shared/config";
+
+// Потенциально, можно передавать accessToken
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/shared/api/typicode/base.ts:4]: -> English

File: todo-app-redux/src/shared/config/index.ts [hunk 17]
```diff
@@ -0,0 +1,27 @@
+/**
+ * Модуль инициализации env-переменных
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/shared/config/index.ts:2]: Please translate all the comments in this file to English

File: todo-app-redux/src/shared/lib/redux-std/root-state.ts [hunk 18]
```diff
@@ -0,0 +1,4 @@
+// RootState намеренно не типизирован что бы скрыть структуру состояния
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/shared/lib/redux-std/root-state.ts:1]: Please translate all the comments in this file to English

File: todo-app-redux/src/shared/lib/redux-std/use-action.ts [hunk 19]
```diff
@@ -0,0 +1,13 @@
+import { useCallback } from "react";
+import { useDispatch } from "react-redux";
+
+export const useAction = <A extends any[]>(
+  // Можно написать более строгие типы если нужно
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/shared/lib/redux-std/use-action.ts:5]: -> English

File: todo-app-redux/src/shared/lib/redux-std/use-action.ts [hunk 20]
```diff
@@ -0,0 +1,13 @@
+import { useCallback } from "react";
+import { useDispatch } from "react-redux";
+
+export const useAction = <A extends any[]>(
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/src/shared/lib/redux-std/use-action.ts:4]: Replace `any` with `unknown` everywhere

File: todo-app-redux/tsconfig.json [hunk 21]
```diff
@@ -0,0 +1,27 @@
+{
+  "compilerOptions": {
+    "baseUrl": "./src",
+    "target": "es5",
```
Comment (Rin-Akaia-eth) [inline todo-app-redux/tsconfig.json:4]: I recommend using `"es6"` target

**Review by Rin-Akaia-eth — CHANGES_REQUESTED** [review]

File: todo-app-redux/src/features/tasks-filters/model/task-filters.slice.ts [hunk 22]
```diff
@@ -0,0 +1,42 @@
+import { createSelector, createSlice, PayloadAction } from "@reduxjs/toolkit";
+import { taskModel } from "entities/task";
+import { createBaseSelector } from "shared/lib/redux-std";
+import { TaskFilters } from "./task-filters.type";
+
+const reducerPatch = "features/task-filters";
```
Comment (nnsense91) [inline todo-app-redux/src/features/tasks-filters/model/task-filters.slice.ts:6]: reducerPath

**Comment by illright** [comment]
Seems like this PR has gone stale, I'll close it

**Pull request closed without merging**